<?xml version="1.0" encoding="utf-8"?>
<robot name="sonar_mexatron">


<material name="gray">
<color rgba="0.5 0.5 0.5 1.0"/>
</material>
<material name="black">
<color rgba="0 0 0 1"/>
</material>
<material name="white">
<color rgba="1 1 1 1"/>
</material>
<material name="gold">
  <color rgba="0.9 0.7 0.2 1.0"/>
</material>

<link name="base_link">
  <inertial>
    <mass value="1.0"/>
    <inertia ixx="0.015" iyy="0.0375" izz="0.0375"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin
      xyz="0 0 0"
      rpy="0 0 1.5708" />
    <geometry>
      <mesh
        filename="package://sonar_mexatron/meshes/chassis_rsz.stl"/>
    </geometry>
<material name="gray"/>
  </visual>
  <collision>
    <origin
      xyz="0 0 0"
      rpy="0 3 0" />
    <geometry>
      <box size='0.15 0.15 0.02'/>
    </geometry>
  </collision>
</link>

<!-- BreadBoard -->
<link name="breadboard">
  <inertial>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <mass
      value="0.001" />
    <inertia
      ixx="0.01"
      ixy="0.01"
      ixz="0.01"
      iyy="0.01"
      iyz="0.01"
      izz="0.01" />
  </inertial>
  <visual>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <geometry>
      <mesh
        filename="package://sonar_mexatron/meshes/breadboard_rsz.stl" />
    </geometry>
<material name="white"/>
  </visual>
  <collision>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <geometry>
      <mesh
        filename="package://sonar_mexatron/meshes/breadboard_rsz.stl" />
    </geometry>
  </collision>
</link>

<joint name="breadboard_to_base" type="fixed">
  <parent link="base_link"/>
  <child link="breadboard"/>
  <origin xyz="0.034 0 0.015" rpy="0 0 1.5708"/>
</joint>

<!-- PCB-->
<link name="pcb_link">
  <inertial>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <mass
      value="0.001" />
    <inertia
      ixx="0.01"
      ixy="0.01"
      ixz="0.01"
      iyy="0.01"
      iyz="0.01"
      izz="0.01" />
  </inertial>
  <visual>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <geometry>
      <mesh
        filename="package://sonar_mexatron/meshes/pcb_rsz.stl" />
    </geometry>
<material name="gold"/>
  </visual>
  <collision>
    <origin
      xyz="0 0 0"
      rpy="0 0 0" />
    <geometry>
      <box size='0.06 0.09 0.01'/>
    </geometry>
  </collision>
</link>

<joint name="pcb_to_base" type="fixed">
  <origin xyz="-0.018 0 0.026" rpy="0 0 1.5708"/>
  <parent link="base_link"/>
  <child link="pcb_link"/>
  <axis xyz="0 0 0"/>
</joint>

<!--Left Wheel-->
  <link
    name="wheellft">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 1.5708 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/wheel_rsz.stl" />
      </geometry>
<material name="black"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <cylinder length="0.01" radius="0.0385"/>
      </geometry>
    </collision>
  </link>

  <joint
    name="wheelleft"
    type="continuous">
    <parent link="base_link" />
    <child link="wheellft" />
    <origin xyz="-0.0345 0.075 -0.012" rpy="-1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

<!--Rigth Wheel-->
  <link
    name="wheelrgt">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 1.5708 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/wheel_rsz.stl" />
      </geometry>
<material name="black"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <cylinder length="0.01" radius="0.0385"/>
      </geometry>
    </collision>
  </link>


  <joint
    name="wheelright"
    type="continuous">
    <parent link="base_link" />
    <child link="wheelrgt" />
    <origin xyz="-0.0345 -0.075 -0.012" rpy="-1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>


<!-- Ball Caster--> 
  <link
    name="caster_link">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
  
    <ode>
      <mu>0</mu>
      <mu2>0</mu2>
      <fdir1> 0 0 0</fdir1>
      <slip1>0</slip1>
      <slip2>0</slip2>
    </ode>
  
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/ballcaster_rsz.stl" />
      </geometry>
<material name="black"/>

    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <sphere radius="0.0125"/>
      </geometry>
    </collision>
  </link>

  <joint
    name="caster_joint"
    type="fixed">
    <origin
        xyz="0.052 0 -0.034"
        rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="caster_link" />
    <axis
      xyz="0 0 1" />
  </joint>

<link
    name="sonar_1">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/sonar.stl" />
      </geometry>
<material name="white"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
  </link>
  
  <joint
    name="sonar_1_joint"
    type="fixed">
    <origin
        xyz="0.052 0 0.03"
        rpy="0 -0.07 0" />
    <parent
      link="base_link" />
    <child
      link="sonar_1" />
    <axis
      xyz="0 0 1" />
  </joint>

<link
    name="sonar_2">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/sonar.stl" />
      </geometry>
<material name="white"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
  </link>
  
  <joint
    name="sonar_2_joint"
    type="fixed">
    <origin
        xyz="0.052 0.05 0.03"
        rpy="0 -0.07 1.2" />
    <parent
      link="base_link" />
    <child
      link="sonar_2" />
    <axis
      xyz="0 0 1" />
  </joint>

<link
    name="sonar_3">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5"
    ixy="0" ixz="0" iyz="0"/>
  </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://sonar_mexatron/meshes/sonar.stl" />
      </geometry>
<material name="white"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
  </link>
  
  <joint
    name="sonar_3_joint"
    type="fixed">
    <origin
        xyz="0.052 -0.05 0.03"
        rpy="0 -0.07 -1.2" />
    <parent
      link="base_link" />
    <child
      link="sonar_3" />
    <axis
      xyz="0 0 1" />
  </joint>


<gazebo>
  <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
    <commandTopic>cmd_vel</commandTopic>
    <odometryTopic>odom</odometryTopic>
    <odometryFrame>odom</odometryFrame>
    <odometrySource>world</odometrySource>
    <publishOdomTF>true</publishOdomTF>
    <robotBaseFrame>base_link</robotBaseFrame>
    <publishWheelTF>false</publishWheelTF>
    <publishTf>true</publishTf>
    <publishWheelJointState>true</publishWheelJointState>
    <legacyMode>false</legacyMode>
    <updateRate>30</updateRate>
    <leftJoint>wheelleft</leftJoint>
    <rightJoint>wheelright</rightJoint>
    <wheelSeparation>0.78</wheelSeparation>
    <wheelDiameter>0.1</wheelDiameter>
    <wheelAcceleration>0.5</wheelAcceleration>
    <wheelTorque>50</wheelTorque>
    <rosDebugLevel>na</rosDebugLevel>
  </plugin>
</gazebo>

<gazebo reference="base_link">
  <material>Gazebo/Grey</material>
</gazebo>
<gazebo reference="breadboard">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="pcb_link">
  <material>Gazebo/Gold</material>
</gazebo>
<gazebo reference="sonar_1">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="sonar_2">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="sonar_3">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="caster_link">
  <material>Gazebo/Black</material>
  <mu1>0</mu1>
  <mu2>0</mu2>
<minDepth>0.001</minDepth>
  <kp>100000000</kp>
  <kd>1</kd>
</gazebo>
<gazebo reference="wheelrgt">
  <material>Gazebo/Black</material>
  <minDepth>0.001</minDepth>
  <kp>100000000</kp>
  <kd>1</kd>
</gazebo>
<gazebo reference="wheellft">
  <material>Gazebo/Black</material>
  <minDepth>0.001</minDepth>
  <kp>100000000</kp>
  <kd>1</kd>
</gazebo>


<gazebo reference="sonar_1">        
   <sensor type="ray" name="front_sonar">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>50</update_rate>
      <ray>
         <scan>
            <horizontal>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.14835</min_angle>
               <max_angle>0.14835</max_angle>
            </horizontal>
            <vertical>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.02</min_angle>
               <max_angle>0.02</max_angle> 
            </vertical>
         </scan>
         <range>
            <min>0.01</min>
            <max>10</max>
            <resolution>0.02</resolution>
         </range>
      </ray>
      <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
         <gaussianNoise>0.005</gaussianNoise>
         <alwaysOn>true</alwaysOn>
         <updateRate>50</updateRate>
         <topicName>front_sonar</topicName>
         <frameName>sonar_1</frameName>
         <radiation>INFRARED</radiation>
         <fov>0.1</fov>
      </plugin>
   </sensor>  
   </gazebo>

<gazebo reference="sonar_2">        
   <sensor type="ray" name="left_sonar">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>50</update_rate>
      <ray>
         <scan>
            <horizontal>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.14835</min_angle>
               <max_angle>0.14835</max_angle>
            </horizontal>
            <vertical>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.02</min_angle>
               <max_angle>0.02</max_angle> 
            </vertical>
         </scan>
         <range>
            <min>0.01</min>
            <max>10</max>
            <resolution>0.02</resolution>
         </range>
      </ray>
      <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
         <gaussianNoise>0.005</gaussianNoise>
         <alwaysOn>true</alwaysOn>
         <updateRate>50</updateRate>
         <topicName>left_sonar</topicName>
         <frameName>sonar_2</frameName>
         <radiation>INFRARED</radiation>
         <fov>0.1</fov>
      </plugin>
   </sensor>  
   </gazebo>

<gazebo reference="sonar_3">        
   <sensor type="ray" name="right_sonar">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>50</update_rate>
      <ray>
         <scan>
            <horizontal>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.14835</min_angle>
               <max_angle>0.14835</max_angle>
            </horizontal>
            <vertical>
               <samples>10</samples>
               <resolution>1</resolution> 
               <min_angle>-0.02</min_angle>
               <max_angle>0.02</max_angle> 
            </vertical>
         </scan>
         <range>
            <min>0.01</min>
            <max>10</max>
            <resolution>0.02</resolution>
         </range>
      </ray>
      <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
         <gaussianNoise>0.005</gaussianNoise>
         <alwaysOn>true</alwaysOn>
         <updateRate>50</updateRate>
         <topicName>right_sonar</topicName>
         <frameName>sonar_3</frameName>
         <radiation>INFRARED</radiation>
         <fov>0.1</fov>
      </plugin>
   </sensor>  
   </gazebo>


</robot>
